library/std/src/time.rs

```diff
@@ -306,11 +283,23 @@ impl Instant {
     /// sleep(Duration::new(1, 0));
     /// let new_now = Instant::now();
     /// println!("{:?}", new_now.duration_since(now));
+    /// println!("{:?}", now.duration_since(new_now)); // 0ns
     /// ```
     #[must_use]
     #[stable(feature = "time2", since = "1.8.0")]
+    #[rustc_inherit_overflow_checks]
+    #[allow(arithmetic_overflow)]
     pub fn duration_since(&self, earlier: Instant) -> Duration {
-        self.0.checked_sub_instant(&earlier.0).expect("supplied instant is later than self")
+        match self.checked_duration_since(earlier) {
+            Some(d) => d,
+            None => {
+                // In debug builds, trigger a panic on None.
+                // This should optimize completely out in release builds.
+                let _ = usize::MAX + 1;
```

Comment:
Presumably the error message produced here is *terrible* -- not saying anything about duration/instant. That feels pretty unfortunate to me. Maybe we should land a compiler intrinsic or similar that basically is a `is_overflow_checks_enabled()` function? That way we can do a normal panic here, with a much better message.

I guess without track_caller we'd at least have the panic originate in library/std/src/time.rs, which seems at least a little better, but maybe indicates we should have a comment on this pointing readers to Instant's monotonic section in the docs so if they look at source code they can figure out what's up fairly quickly.

Reply:
At the very least, we could rename the constant, so that the error message contains a hint (and a string that's easy to search for).

```suggestion
                const DURATION_SINCE_ON_FUTURE_TIME: usize = usize::MAX;
                let _ = DURATION_SINCE_ON_FUTURE_TIME + 1;
```